Adding Content

esco opened the pull request:
Assumptions:

Student knows what primitives are after learning about variables
Student already knows what enumeration is after learning about Arrays
Student knows what new lines are after learning about strings
Student learned about scopes in the lesson about functions 
Student learned about functions but not prototype so I have a quick intro to prototype

pletcher commented:
Hi @esco, thanks for getting this in so quickly! Overall, this is really solid. The main bit of feedback that I have is that at times, it gets a bit too technical for beginners: students don't need to know (yet, anyway) about JavaScript's GC behavior.

I'll add more detailed comments inline.

Thanks again,
Chuck

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+var pikachu = { "name": "Pikachu", "type": "Electric", "species": "Mouse", "level": 100, "attacks": ["Thunderbolt", "Quick Attack", "Growl", "Slam"], "speak": function(){ return "pika pika"}};
+```
+
+Here we have something that looks similar to an Array in that we have values that are comma seperated, but there are a few differences:
````
Nit: seperated -> separated

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+- Instead of using square brackets `[]` we're using curly braces `{}`.
+
+- Between the commas, we have two pieces of information instead of one. These are known as `key-value` pairs. Keys are strings that describe a value and are followed by a colon `:`. The colon is followed by the value that the key represents. In `"name": "Pikachu"`, "name" is the `key` and "Pikachu" is the `value`.
```
I really like how you introduce students to kv pairs

esco replied:
Thanks!

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+var pikachu  = ["Pikachu", "Electric", "Mouse",  100, "Thunderbolt", "Quick Attack", "Growl", "Slam", function speak(){return "pika pika";}]
+```
+
+This gives us a collection of data about Pikachu, but accessing specific information is difficult and messy. If we want to know Pikachu's name, type, species, level..etc we'd need to know which index in the array to use.
````
Nit: "name, type, species, level..etc" -> "name, type, species, level, etc."

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+};
+```
+
+This is much more readable than the Array. We know exactly what each value on the object represents because each is associated with a descriptive key. Instead of using Array index numbers, we reference values using the key strings. If we wanted to get Pikachu's name,type,species,level..etc we don't have to remember some random number.
````
Nit: "name,type,species,level..etc" -> "name, type, species, level, etc."

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+`pikachu["level"]` is much easier to understand than `pikachu[3]` because the key gives us some context.
+
+Each key on the Object is also called a `property` of the Object. An Object's properties can also be accessed using dot notation:
```
👍

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+pikachu["speak"](); // "pika pika"
+```
+
+Notice how we stored many properties of different types on the pikachu Object. Objects are like containers for storing variables. An Object can store any type of variable such as a primitive, functions or even other Objects. Objects have many uses besides representing a single entity such as a Pokemon. The main idea behind Objects is that they provide a way to group variables that are easily acce ...
````
This is great

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+
+### Primer on Prototypes
+Throughout this lesson you'll see the terms `prototype` and `prototype chain` a few times. The prototype is like DNA. Its passed down from parent to child, to grandchild...etc. The prototype chain refers to the list of all ancestors (parents, grandparents, great grandparents...etc). Just as a child would inherit properties from their parents, Objects and other JavaScript types inherit properties  ...
```
"parents, grandparents, great grandparents...etc" -> "parents, grandparents, great grandparents, etc."

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+#### Using Object.create method
+
+`Object.create` is a method for creating Objects with a specific prototype. Passing `null` will create an Object that has no prototype. This can be useful when we don't care about inherting properties from the prototype chain.
```
This might be a little advanced for students at this stage, simply because they haven't practiced working with prototypes much yet. Is there a gentler way to ease them into `Object.create()`?

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+## Adding properties to Objects
+
+By default, Objects are mutable. This means that after an Object is created, we can add additional properties to it at any time. There are a few ways to add a property to an Object:
```
Great!

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+};
+```
+
+When using this syntax, keys don't have to be wrapped in quotes. The following is also valid:
````
Would it make sense to clarify this w/r/t https://github.com/learn-co-curriculum/javascript-curriculum-assignment-objects-readme/pull/4/files#diff-04c6e90faac2675aa89e2176d2eec7d8R164 ?

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+```js
+var prop = "name";
+
+pikachu[prop] = "Pikachu"; // same as pikachu["name"] = "Pikachu"
````
👍

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+This is invalid because variable names can't have hyphens. If we wanted to use a key that doesn't follow JavaScript's naming constraints for variables, we'd have to use the key index since strings can hold any character:
+
+```js
+pikachu["max-level"] = 100; // this is valid
````
👍

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+pikachu["max-level"] = 100; // this is valid
+```
+
+#### Object.defineProperty
````
This section down to the section on `Object.freeze` could, I think, be broken out into a separate lesson (or separate lessons).

esco replied:
Agreed, its a lot to take in. I'll remove it from this lesson.

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+`pikachu.level` doesn't contain a value, instead it uses `pikachu._level` to store and retrieve a value that is set within the range [1, 100].
+
+<sub>Side Note: Its still possible to modify `pikachu._level` directly. In order to truly get the benefit, we'd have to define the `_level` variable in a private scope where it can't be modified outside of this Object.</sub>
```
👍

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+## Accessing Object Properties
+
+Earlier we noted how there were two ways of assigning a property to an Object that resembled variable assignment. We can access these assigned properties using a similar approach. We've actually seen this numerous times throghout the lesson:
```
"throghout" -> "throughout"

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+## Removing Object Properties
+
+Sometimes you'll want to remove a property from an Object when its no longer necessary. This can help free up memory by removing references to variables that are no longer needed. Since JavaScript is a memory managed language theres no guarantee exactly when memory will be freed, but you can at least ensure that it will happen some time in the future by removing all unused references.
```
I think this section is great, but we can probably cut the part about memory management -- it's beyond the scope of what students will need to practice in the near term (although at some point they should understand it, for sure).

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+
+#### for-in loop
+
+A for-in loop allows you to iterate through an Object's keys. As part of the for-in loop, we declare a variable and provide an Object to iterate through. In the example below we're going to iterate through our coupons, storing each key in a new variable we create as part of the loop called `code`.
```
👍

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+```js
+for (var code in coupons) {
+	if (!coupons.hasOwnProperty(code)) {
+	 	break;
````
We want `continue` here so that it behaves as described below, right? ("Adding this check will skip any property that is not a direct property of the Object.") As written, this will skip every property after the first property that isn't an own property of Object.

esco replied:
good catch!

pletcher commented on README.md:
```diff
@@ -1,20 +1,453 @@
+Adding this check will skip any property that is not a direct property of the Object.
+
+
+Another way to find out what keys an Object has is using the `Object.keys` method. This method returns an array of key strings that the Object contains. `Object.keys` only returns direct properties of the Object. Use of `Object.hasOwnProperty` isn't necessary.
```
👍

pletcher commented on README.md:
````diff
@@ -1,20 +1,453 @@
+Object.keys(coupons) // ["QHTYB", "HOLIDAY", "MOVIE20"]
+```
+
+Again, since Objects are unordered, the order of the keys in the returned Array is not guarenteed to match the order in which they were added to the Object.
````
"guarenteed" -> "guaranteed"

esco commented:
Thanks for the feedback. I'll review tonight.

esco commented:
@pletcher sorry about all the typos! I enabled spell check for markdown files in my editor so this shouldn’t be an issue going forward 👍 

Changes:
- Fixed typos
- replaced `break` with `continue`
- Removed `Object.defineProperty`, `Object.freeze`, `Object.seal` and also removed references to the writable and configurable descriptors in the “Removing Object Properties” section.
- Added a gentler explanation of `Object.create`
- Added section on valid property names to include both dot and bracket notation
- Removed references to garbage collection
- Renamed Key Indexes —> Bracket Notation
